### PR TITLE
Send non-printable codepoints as U+FFFD

### PR DESCRIPTION
**Pull request by adsr**
Presently if someone sets a cell to an invalid codepoint, termbox will write it out verbatim, but it's guaranteed to be garbage. Instead send U+FFFD. This effectively limits cell contents to valid Unicode.

~Any reasons to not do this? Could we be even more strict and use `iswprint`?~

EDIT: Updated PR to use `iswprint`

Note users can still send arbitrary data via `tb_send`.

### CODE REVIEW

File: termbox2.h [hunk 1]
```diff
@@ -1535,6 +1535,7 @@ static int send_sgr(uint32_t fg, uint32_t bg, int fg_is_default,
 static int send_cursor_if(int x, int y);
 static int send_char(int x, int y, uint32_t ch);
 static int send_cluster(int x, int y, uint32_t *ch, size_t nch);
+static int is_valid_codepoint(uint32_t ch);
```
Comment (pepe) [inline termbox2.h:1538]: Just yak shaving, but  `is_invalid_codepoint` with inverted logic will be more suitable here. No ternary operator and `not`ing its return :-).

I reached this conclusion when I read the function definition and encountered the ternary in the `return`. And it is entirely possible you did it intentionally, in which case, ignore this.

Comment (pepe) [reply termbox2.h:1538]: The word `invalid` is even in the title of the PR :-D.

**Comment by adsr** [comment]
Thank you for the review @pepe.

I altered the PR to filter based on `iswprint` instead of `is_valid_codepoint`. Is this too strict? I think it's OK.

| chars (hex range) | description | iswprint |
| -- | -- | -- |
| 00-08 | control codes | no |
| 09-0d | whitespace | no |
| 0e-1f | control codes | no |
| 20-7e | printable ASCII | yes |
| 7f | delete | no |

Of course there are codepoints above the 1-byte range, but the same categories apply. I think it comes down to control codes and whitespaces.

* Control codes: I can't think of any use cases for setting a cell to a control code. Am I missing anything? I'll re-emphasize that users can send arbitrary data via `tb_send*`. (Although sending many of these control codes with termbox still active will probably break rendering.)
* Whitespaces: I could see users wanting to use newlines specifically in `tb_print*`. I have an open PR to support that [here](https://github.com/termbox/termbox2/pull/77). IMO it's not worth supporting tab, form feed (`\f`), vertical tab (`\v`), or macOS- or Windows-style newlines.

Let me know if you can think of any counter-examples. I will leave these PRs open for a while for visibility.

**Comment by pepe** [comment]
Thank you. That is a much better choice, anyway. I am planning to rewrite a TUI of mine in the coming weeks, so I will definitely use this version to test it.

**Pull request merged**